src/potdata/samplers.py

```diff
@@ -616,3 +616,245 @@ def plot2(self, separation1: int, separation2: int, show: bool = False,
+
+    @requires(
+        SOAP,
+        "`dscribe` is required to use the sampler. To install it, see "
+        "https://github.com/SINGROUP/dscribe",
+    )
+    # This can be a staticmethod; not use because @requires does not work with it
+    def _get_soap_vector_atom(
+        self, data: list[Structure], soap_kwargs: dict
+    ) -> list[np.ndarray]:
+        """Convert structures to SOAP vectors of all atoms.
+
+        Returns:
+            SOAP vectors for all structures. For each structure, the SOAP vectors is a
+            2D array of shape (n_atoms, n_features), where n_atoms is the number of
+            atoms and n_features is the number of features in the SOAP vector.
+        """
+
+        species = set()
+        for structure in data:
+            species.update(structure.symbol_set)
+
+        if "species" in soap_kwargs:
+            raise ValueError(
+                "The `species` argument in `soap_kwargs` is not allowed. "
+                "The species will be automatically determined from the structures."
+            )
+
+        soap = SOAP(species=species, **soap_kwargs)
+        atoms = [AseAtomsAdaptor.get_atoms(structure) for structure in data]
+        soap_vectors = list(soap.create(atoms))
+        
+        return soap_vectors
+
+    @staticmethod
+    def _select_by_species(
+        structures: list[Structure], vectors: list[np.ndarray], species: list[str]
+    ) -> list[np.ndarray]:
+        """Select SOAP vectors for atoms of specific species."""
+
+        def select_one(struct, vec):
+            indices = [i for i, s in enumerate(struct.species) if s.symbol in species]
+            return vec[indices]
+
+        selected = []
+        for struct, vec in zip(structures, vectors):
+            selected.append(select_one(struct, vec))
+
+        return selected
+
+    @staticmethod
+    def _get_soap_vector_structure(vectors: list[np.ndarray], pool: str) -> np.ndarray:
+        """Convert SOAP vectors of atoms to SOAP vectors of structures.
+
+        Return a 2D array of shape (n_structures, n_features).
+        """
+        if pool.lower() == "mean":
+            return np.asarray([np.mean(x, axis=0) for x in vectors])
+        elif pool.lower() == "concatenate":
+            return np.asarray([np.ravel(x) for x in vectors])
+        else:
+            supported = ["mean", "concatenate"]
+            raise ValueError(
+                f"Unsupported pooling method `{pool}`. Expected one of " f"{supported}."
+            )
+
+    @staticmethod
+    def _dim_reduction(x: np.ndarray, dim: int | float):
+        """Perform dimension reduction on a 2D array."""
+        pca = PCA(n_components=dim)
+        reduced = pca.fit_transform(x)
+
+        return reduced
+
+    def _cluster_kmeans(self, data: np.ndarray) -> list[int]:
+        """
+        Perform KMeans clustering and return cluster labels.
+        """
+
+        clustering = KMeans(n_clusters=self.kmeans_kwargs["n_clusters"], n_init=10).fit(data)
```

Comment:
You should not hard-code `n_init=10` here. Remove it. Take a look at how we do it in line 394. Basically, this line should be 
```clustering = KMeans(**self.kmeans_kwargs).fit(data)```
which will provide all args you've set in `__init__` to KMeans